### PR TITLE
[Merged by Bors] - feat(List/MinMax): maximum_of_length_pos

### PR DESCRIPTION
**Pull request by kim-em**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/MinMax.lean [hunk 1]
```diff
@@ -392,25 +392,66 @@ theorem coe_le_maximum_iff : a ≤ l.maximum ↔ ∃ b, b ∈ l ∧ a ≤ b := b
...
-    simp [List.maximum_cons, ih]
+    simp [maximum_cons, ih]
 
-theorem minimum_le_coe_iff : l.minimum ≤ a ↔ ∃ b, b ∈ l ∧ b ≤ a := by
-  induction l with
-  | nil => simp
-  | cons h t ih =>
-    simp [List.minimum_cons, ih]
+theorem minimum_le_coe_iff : l.minimum ≤ a ↔ ∃ b, b ∈ l ∧ b ≤ a :=
+  @coe_le_maximum_iff αᵒᵈ _ _ _
```
Comment (ocfnash) [inline Mathlib/Data/List/MinMax.lean:398]:
```suggestion
  coe_le_maximum_iff (α := αᵒᵈ)
```

File: Mathlib/Data/List/MinMax.lean [hunk 2]
```diff
@@ -392,25 +392,66 @@ theorem coe_le_maximum_iff : a ≤ l.maximum ↔ ∃ b, b ∈ l ∧ a ≤ b := b
...
-  match l, h with | _ :: _, _ => by simp [List.maximum_cons]
+  match l, h with | _ :: _, _ => by simp [maximum_cons]
 
 theorem minimum_ne_top_of_length_pos {l : List α} (h : 0 < l.length) : l.minimum ≠ ⊤ :=
-  @List.maximum_ne_bot_of_length_pos αᵒᵈ _ _ h
+  @maximum_ne_bot_of_length_pos αᵒᵈ _ _ h
+
+/-- The maximum value in a non-empty `List`. -/
+def maximum_of_length_pos {L : List α} (h : 0 < L.length) : α :=
+  WithBot.unbot L.maximum (maximum_ne_bot_of_length_pos h)
```
Comment (ocfnash) [inline Mathlib/Data/List/MinMax.lean:414]: We already have `{l : List α}` from `variable` on line 342 so we can drop a lot of `{L : List α}` and `{l : List α}` below:
```suggestion
def maximum_of_length_pos (h : 0 < l.length) : α :=
  WithBot.unbot l.maximum (maximum_ne_bot_of_length_pos h)
```

File: Mathlib/Data/List/MinMax.lean [hunk 3]
```diff
@@ -392,25 +392,66 @@ theorem coe_le_maximum_iff : a ≤ l.maximum ↔ ∃ b, b ∈ l ∧ a ≤ b := b
...
+    (L.minimum_of_length_pos h : α) = L.minimum :=
+  WithTop.coe_untop _ _
+
+theorem le_maximum_of_length_pos_iff {L : List α} (h : 0 < L.length) :
+    b ≤ maximum_of_length_pos h ↔ b ≤ L.maximum :=
+  WithBot.le_unbot_iff _
+
+theorem minimum_of_length_pos_le_iff {L : List α} (h : 0 < L.length) :
+    minimum_of_length_pos h ≤ b ↔ L.minimum ≤ b :=
+  @le_maximum_of_length_pos_iff αᵒᵈ _ _ _ h
```
Comment (ocfnash) [inline Mathlib/Data/List/MinMax.lean:436]:
```suggestion
  le_maximum_of_length_pos_iff (α := αᵒᵈ) h
```

File: Mathlib/Data/List/MinMax.lean [hunk 4]
```diff
@@ -392,25 +392,66 @@ theorem coe_le_maximum_iff : a ≤ l.maximum ↔ ∃ b, b ∈ l ∧ a ≤ b := b
...
+
+theorem le_maximum_of_length_pos_of_mem {L : List α} (h : a ∈ L) (w : 0 < L.length) :
+     a ≤ L.maximum_of_length_pos w := by
+  simp [le_maximum_of_length_pos_iff]
+  exact le_maximum_of_mem' h
+
+theorem minimum_of_length_pos_le_of_mem {L : List α} (h : a ∈ L) (w : 0 < L.length) :
+     L.minimum_of_length_pos w ≤ a :=
+  @le_maximum_of_length_pos_of_mem αᵒᵈ _ _ _ h _
```
Comment (ocfnash) [inline Mathlib/Data/List/MinMax.lean:445]:
```suggestion
  le_maximum_of_length_pos_of_mem (α := αᵒᵈ) h w
```

File: Mathlib/Data/List/MinMax.lean [hunk 5]
```diff
@@ -392,25 +392,66 @@ theorem coe_le_maximum_iff : a ≤ l.maximum ↔ ∃ b, b ∈ l ∧ a ≤ b := b
...
+
+theorem getElem_le_maximum_of_length_pos {L : List α} (w : i < L.length) :
+    L[i] ≤ L.maximum_of_length_pos (Nat.zero_lt_of_lt w) := by
+  apply le_maximum_of_length_pos_of_mem
+  exact get_mem L i w
+
+theorem minimum_of_length_pos_le_getElem {L : List α} (w : i < L.length) :
+    L.minimum_of_length_pos (Nat.zero_lt_of_lt w) ≤ L[i] :=
+  @getElem_le_maximum_of_length_pos αᵒᵈ _ _ _ _
```
Comment (ocfnash) [inline Mathlib/Data/List/MinMax.lean:454]:
```suggestion
  getElem_le_maximum_of_length_pos (α := αᵒᵈ) w
```

File: Mathlib/Data/List/MinMax.lean [hunk 6]
```diff
@@ -392,25 +392,66 @@ theorem coe_le_maximum_iff : a ≤ l.maximum ↔ ∃ b, b ∈ l ∧ a ≤ b := b
...
-  @List.maximum_ne_bot_of_ne_nil αᵒᵈ _ _ h
+  @maximum_ne_bot_of_ne_nil αᵒᵈ _ _ h
 
 theorem maximum_ne_bot_of_length_pos {l : List α} (h : 0 < l.length) : l.maximum ≠ ⊥ :=
-  match l, h with | _ :: _, _ => by simp [List.maximum_cons]
+  match l, h with | _ :: _, _ => by simp [maximum_cons]
 
 theorem minimum_ne_top_of_length_pos {l : List α} (h : 0 < l.length) : l.minimum ≠ ⊤ :=
-  @List.maximum_ne_bot_of_length_pos αᵒᵈ _ _ h
+  @maximum_ne_bot_of_length_pos αᵒᵈ _ _ h
```
Comment (ocfnash) [inline Mathlib/Data/List/MinMax.lean:410]:
```suggestion
  maximum_ne_bot_of_length_pos (α := αᵒᵈ) h
```

File: Mathlib/Data/List/MinMax.lean [hunk 7]
```diff
@@ -392,25 +392,66 @@ theorem coe_le_maximum_iff : a ≤ l.maximum ↔ ∃ b, b ∈ l ∧ a ≤ b := b
...
+
+/-- The maximum value in a non-empty `List`. -/
+def maximum_of_length_pos {L : List α} (h : 0 < L.length) : α :=
+  WithBot.unbot L.maximum (maximum_ne_bot_of_length_pos h)
+
+/-- The minimum value in a non-empty `List`. -/
+def minimum_of_length_pos {L : List α} (h : 0 < L.length) : α :=
+  @maximum_of_length_pos αᵒᵈ _ _ h
```
Comment (ocfnash) [inline Mathlib/Data/List/MinMax.lean:418]:
```suggestion
  maximum_of_length_pos (α := αᵒᵈ) h
```

File: Mathlib/Data/List/MinMax.lean [hunk 8]
```diff
@@ -392,25 +392,66 @@ theorem coe_le_maximum_iff : a ≤ l.maximum ↔ ∃ b, b ∈ l ∧ a ≤ b := b
...
+lemma coe_maximum_of_length_pos {L : List α} (h : 0 < L.length) :
+    (L.maximum_of_length_pos h : α) = L.maximum :=
+  WithBot.coe_unbot _ _
+
+@[simp]
+lemma coe_minimum_of_length_pos {L : List α} (h : 0 < L.length) :
+    (L.minimum_of_length_pos h : α) = L.minimum :=
+  WithTop.coe_untop _ _
+
+theorem le_maximum_of_length_pos_iff {L : List α} (h : 0 < L.length) :
```
Comment (ocfnash) [inline Mathlib/Data/List/MinMax.lean:430]:
```suggestion
@[simp]
theorem le_maximum_of_length_pos_iff {L : List α} (h : 0 < L.length) :
```

File: Mathlib/Data/List/MinMax.lean [hunk 9]
```diff
@@ -392,25 +392,66 @@ theorem coe_le_maximum_iff : a ≤ l.maximum ↔ ∃ b, b ∈ l ∧ a ≤ b := b
...
+    b ≤ maximum_of_length_pos h ↔ b ≤ L.maximum :=
+  WithBot.le_unbot_iff _
+
+theorem minimum_of_length_pos_le_iff {L : List α} (h : 0 < L.length) :
```
Comment (ocfnash) [inline Mathlib/Data/List/MinMax.lean:434]:
```suggestion
@[simp]
theorem minimum_of_length_pos_le_iff {L : List α} (h : 0 < L.length) :
```

File: Mathlib/Data/List/MinMax.lean [hunk 10]
```diff
@@ -392,25 +392,66 @@ theorem coe_le_maximum_iff : a ≤ l.maximum ↔ ∃ b, b ∈ l ∧ a ≤ b := b
...
+     a ≤ L.maximum_of_length_pos w := by
+  simp [le_maximum_of_length_pos_iff]
+  exact le_maximum_of_mem' h
+
+theorem minimum_of_length_pos_le_of_mem {L : List α} (h : a ∈ L) (w : 0 < L.length) :
+     L.minimum_of_length_pos w ≤ a :=
+  @le_maximum_of_length_pos_of_mem αᵒᵈ _ _ _ h _
+
+theorem getElem_le_maximum_of_length_pos {L : List α} (w : i < L.length) :
+    L[i] ≤ L.maximum_of_length_pos (Nat.zero_lt_of_lt w) := by
```
Comment (ocfnash) [inline Mathlib/Data/List/MinMax.lean:448]: I'm not actually sure but I think an `optParam` might make this more useful:
```suggestion
theorem getElem_le_maximum_of_length_pos {L : List α} (w : i < L.length)
    (h := (Nat.zero_lt_of_lt w)) :
    L[i] ≤ L.maximum_of_length_pos h := by
```
(likewise for `minimum_of_length_pos_le_getElem` below)

**Review by ocfnash — APPROVED** [review]
Thanks!

bors d+

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**